core/src/main/java/io/appflate/restmock/utils/RestMockUtils.java

```diff
@@ -35,4 +42,31 @@ public static MockResponse createResponseFromFile(RESTMockFileParser RESTMockFil
         String fileContents = RESTMockFileParser.readJsonFile(jsonFilePath);
         return new MockResponse().setResponseCode(responseCode).setBody(fileContents);
     }
+
+    /**
+     * Extract query parameters from a {@link URL}.
+     *
+     * Note: Gratuitously stolen from http://stackoverflow.com/a/13592567/281460
+     *
+     * @param url The {@link URL} to retrieve query parameters for.
+     *
+     * @return A {@link Map} of {@link String} key / a {@link List} of {@link String} value pairs
+     *         appearing in the query parameters of the given URL. This will never be null.
+     *
+     * @throws UnsupportedEncodingException If unable to decode from UTF-8. This should never happen.
+     */
+    public static Map<String, List<String>> splitQuery(URL url) throws UnsupportedEncodingException {
+      final Map<String, List<String>> query_pairs = new LinkedHashMap<String, List<String>>();
+      final String[] pairs = url.getQuery().split("&");
+      for (String pair : pairs) {
+        final int idx = pair.indexOf("=");
+        final String key = idx > 0 ? URLDecoder.decode(pair.substring(0, idx), "UTF-8") : pair;
+        if (!query_pairs.containsKey(key)) {
+          query_pairs.put(key, new LinkedList<String>());
+        }
```

Comment:
if we plan to support multiple values for a single query param, then it would be nice to extend the `QueryParam` class for that by adding a constructor `QueryParam(String key, String... values)` and storing the values in a list. This way the `hasExactQueryParameters` method will really check if the given queryParams EXACTLY matches the ones in the url